falcon/media/urlencoded.py

```diff
@@ -0,0 +1,45 @@
+from urllib.parse import urlencode
+
+from falcon.media.base import BaseHandler
+from falcon.util.uri import parse_query_string
+
+
+class URLEncodedFormHandler(BaseHandler):
+    """
+    URL-encoded form data handler.
+
+    This handler parses ``application/x-www-form-urlencoded`` HTML forms to a
+    ``dict`` in a similar way that URL query parameters are parsed.
+
+    Keyword Arguments:
+        keep_blank (bool): Whether to keep empty-string values from the form
+            when deserializing.
+        csv (bool): Whether to split comma-separated form values into list
+            when deserializing.
+    """
+
+    def __init__(self, keep_blank=True, csv=False):
+        self.keep_blank = keep_blank
+        self.csv = csv
+
+    def serialize(self, media, content_type):
+        # NOTE(vytas): Setting doseq to True to mirror the parse_query_string
+        # behaviour.
+        return urlencode(media, doseq=True)
+
+    def deserialize(self, stream, content_type, content_length):
+        body = stream.read()
+
+        # NOTE(kgriffs): According to http://goo.gl/6rlcux the
+        # body should be US-ASCII. Enforcing this also helps
+        # catch malicious input.
+        body = body.decode('ascii')
+
+        # NOTE(vytas): Due to the Request.media property mechanics, this path
+        # is normally never hit.
+        if not body:
+            return None
```

Comment:
Would it be more type-correct to return `{}`?

Reply:
No, honestly, I think I disagree an empty dict is any more correct. JSON media can also be a JSON list, a JSON string or any other JSON-serializable document. Unfortunately I don't have a one-size-fits-all suggestion...
See also the discussion on the same matter here: https://github.com/falconry/falcon/issues/1216

In the code highlighted above, though, I am just cementing the fact how Falcon processes media now. As stated in the comment, this path is normally never hit when using this (or other) handler due to it being short-circuited before the handler even gets a chance to see the request media.

Reply:
I'm actually thinking to go ahead and create an issue for this.
You're in principle right @ch3pjw -- an empty URL-encoded form string should deserialize to an empty dict (while I do *not* think it should be the default behaviour for JSON).

Reply:
@vytas7 If we are to proceed as I proposed in #1589, you could just go ahead and make the change now to return `{}`.

Reply:
Will do.
Another alternative is just to remove this check altogether and let the `parse_query_string` handle this, considering it is not a very common case for submitted media?

Reply:
Good point. Letting `parse_query_string()` handle it would require a few more CPU cycles, but given that it should not be a common case, the extra overhead isn't worth bothering about.

Reply:
Uh oh. I'll have to short circuit for now anyway to keep the things moving... :grimacing: 
And file this as a new issue I guess:
```python
>>> from urllib import parse
>>> from falcon.util import uri
>>> parse.parse_qs('', keep_blank_values=True)
{}
>>> uri.parse_query_string('', keep_blank=True)
{'': ''}
```